src/java/org/apache/cassandra/db/compaction/UnifiedCompactionStrategy.java

```diff
@@ -388,6 +444,73 @@ public UnifiedCompactionTask createCompactionTask(LifecycleTransaction txn, fina
         return createCompactionTask(txn, gcBefore);
     }
 
+    private Collection<CompactionTask> createParallelCompactionTasks(LifecycleTransaction transaction, int gcBefore)
+    {
+        // Separating the expired sstables is too hard, just use all (C* 5's version of UCS splits these out and will solve this).
+        Collection<SSTableReader> sstables = transaction.originals();
+        ShardManager shardManager = getShardManager();
+        double density = shardManager.calculateCombinedDensity(sstables);
+        int numShards = controller.getNumShards(density * shardManager.shardSetCoverage());
+        if (numShards <= 1)
+            return Collections.singletonList(createCompactionTask(transaction, gcBefore));
+
+        CompositeLifecycleTransaction compositeTransaction = new CompositeLifecycleTransaction(transaction);
+        List<CompactionTask> tasks = splitSSTablesInShards(sstables,
+                                                           shardManager.boundaries(numShards),
+                                                           (rangeSSTables, range) -> new UnifiedCompactionTask(realm,
+                                                                                                                   this,
+                                                                                                                   new PartialLifecycleTransaction(compositeTransaction),
+                                                                                                                   gcBefore,
+                                                                                                                   shardManager,
+                                                                                                                   range,
+                                                                                                                   rangeSSTables));
+
+        if (tasks.isEmpty())
+            transaction.close(); // this should not be reachable normally, close the transaction for safety
+
+        if (tasks.size() == 1) // if there's just one range, make it a non-ranged task (to apply early open etc.)
+        {
+            assert tasks.get(0).inputSSTables().equals(transaction.originals());
+            return Collections.singletonList(createCompactionTask(transaction, gcBefore));
+        }
+        else
+            return tasks;
+    }
+
+    private <T> List<T> splitSSTablesInShards(Collection<SSTableReader> sstables,
```

Comment:
what about making this method static and writing specific unit tests to cover all of the cases?